src/eradiate_plugins/tests/bsdfs/test_hapke.py

```diff
@@ -0,0 +1,333 @@
+import drjit as dr
+import mitsuba as mi
+import numpy as np
+import xarray as xr
+import pytest
+
+from os.path import join
+
+import matplotlib.pyplot as plt
+
+from ..tools import sample_eval_pdf_bsdf
+from mitsuba.scalar_rgb.test.util import find_resource
+
+
+@pytest.fixture
+def static_reference():
+
+    references = find_resource("resources/tests/eradiate_plugins/bsdfs")
+
+    hapke_reference_filename = join(references, "hapke_principal_plane_example.nc")
+
+    return xr.load_dataset(hapke_reference_filename)
+
+
+def test_create_hapke(variant_scalar_rgb):
+    # Test constructor of 3-parameter version of RPV
+    rtls = mi.load_dict({"type": "hapke"})
+
+    assert isinstance(rtls, mi.BSDF)
+    assert rtls.component_count() == 1
+    assert rtls.flags(0) == mi.BSDFFlags.GlossyReflection | mi.BSDFFlags.FrontSide
+    assert rtls.flags() == rtls.flags(0)
+
+    params = mi.traverse(rtls)
+    assert "w.value" in params
+    assert "b.value" in params
+    assert "c.value" in params
+    assert "theta.value" in params
+    assert "B_0.value" in params
+    assert "h.value" in params
+
+
+def angles_to_directions(theta, phi):
+    return mi.Vector3f(
+        np.sin(theta) * np.cos(phi), np.sin(theta) * np.sin(phi), np.cos(theta)
+    )
+
+
+def eval_bsdf(bsdf, wi, wo):
+    si = mi.SurfaceInteraction3f()
+    si.wi = wi
+    ctx = mi.BSDFContext()
+    return bsdf.eval(ctx, si, wo, True)[0]
+
+
+def test_defaults_and_print(variant_scalar_mono):
+    rtls = mi.load_dict({"type": "rtls"})
+    value = str(rtls)
+    reference = "\n".join(
+        [
+            "RTLSBSDF[",
+            "  f_iso = UniformSpectrum[value=0.209741],",
+            "  f_vol = UniformSpectrum[value=0.081384],",
+            "  f_geo = UniformSpectrum[value=0.004140],",
+            "  h = 2,",
+            "  r = 1,",
+            "  b = 1",
+            "]",
+        ]
+    )
+    assert reference == value
+
+
+regression_test_geometries = [
+    (0.18430089, 1.46592582, 3.92553451, 0.39280281),
+]
+
+
+
+def angles_to_directions(theta, phi):
+    return mi.Vector3f(
+        np.sin(theta) * np.cos(phi), np.sin(theta) * np.sin(phi), np.cos(theta)
+    )
+
+def test_hapke_hotspot(variant_scalar_mono):
+    c = 0.273
+    c = (1 + c) / 2
+
+    hapke = mi.load_dict({ # Pommerol et al. (2013)
+        "type": "hapke",
+        "w": 0.526,
+        "theta": 13.3,
+        "b": 0.187,
+        "c": c,
+        "h": 0.083,
+        "B_0": 1.0,
+    })
+
+    theta_i = np.deg2rad(30.0)
+
+    theta_o = np.deg2rad(30.0)
+
+    phi_i = np.deg2rad(0.0)
+
+    phi_o = np.deg2rad(0.0)
+
+    wi = angles_to_directions(theta_i, phi_i)
+    wo = angles_to_directions(theta_o, phi_o)
+
+    values = np.asarray(eval_bsdf(hapke, wi, wo)) / np.abs(np.cos(theta_o))
+    assert np.allclose(values, 0.24746648)
+
+
+def test_hapke_grazing_viewing_angle(variant_scalar_mono):
```

Comment:
Important: when dealing with BSDFs, there is no such thing as "viewing" angles, only incoming and outgoing directions.

```suggestion
def test_eval_grazing_outgoing_direction(variant_scalar_mono):
```